server/src-lib/Hasura/RQL/DDL/Schema.hs

```diff
@@ -58,22 +61,26 @@ data RunSQL
   { rSql                      :: Text
   , rCascade                  :: !(Maybe Bool)
   , rCheckMetadataConsistency :: !(Maybe Bool)
+  , rReadOnly                 :: !(Maybe Bool)
   } deriving (Show, Eq, Lift)
 $(deriveJSON (aesonDrop 1 snakeCase){omitNothingFields=True} ''RunSQL)
 
 runRunSQL :: (CacheBuildM m, UserInfoM m) => RunSQL -> m EncJSON
-runRunSQL (RunSQL t cascade mChkMDCnstcy) = do
+runRunSQL (RunSQL sql cascade mChkMDCnstcy (fromMaybe False -> isReadOnly)) = do
   adminOnly
-  isMDChkNeeded <- maybe (isAltrDropReplace t) return mChkMDCnstcy
-  bool (execRawSQL t) (withMetadataCheck (or cascade) $ execRawSQL t) isMDChkNeeded
+  isMDChkNeeded <- maybe (if isReadOnly then pure False else isAltrDropReplace sql) return mChkMDCnstcy
+  bool (execRawSQL sql) (withMetadataCheck (or cascade) $ execRawSQL sql) isMDChkNeeded
   where
     execRawSQL :: (MonadTx m) => Text -> m EncJSON
     execRawSQL =
-      fmap (encJFromJValue @RunSQLRes) . liftTx . Q.multiQE rawSqlErrHandler . Q.fromText
+      fmap (encJFromJValue @RunSQLRes) . liftTx . setTransactionAccess . Q.multiQE rawSqlErrHandler . Q.fromText
       where
         rawSqlErrHandler txe =
           let e = err400 PostgresError "query execution failed"
           in e {qeInternal = Just $ toJSON txe}
+    setTransactionAccess tx = if isReadOnly then setReadOnly >> tx else tx
+
+    setReadOnly =  Q.unitQE defaultTxErrorHandler "SET TRANSACTION READ ONLY" () False
```

Comment:
One issue I foresee with this is using read-only `run_sql` queries in `bulk` queries, since we run entire `bulk` queries in one transaction, so setting the transaction to read-only will cause problems for other queries that come later if they made modifications.

Personally, I think it would be nice to add some sort of scoped `withAccessMode` abstraction somewhere in  `Hasura/SQL/` with a type like

```haskell
withAccessMode :: (MonadTx m) => TxAccess -> m a -> m a
```

that gets the current access mode, sets it if necessary, and then resets it to the old value when control leaves the function. (Using some form of lifted `bracket` operation is probably the right thing to implement `withAccessMode`, which would additionally require something like `MonadMask` or `MonadBaseControl`.)

Reply:
ah right! Will incorporate this.

Reply:
TBH, I have no idea how to unset a transaction. Do you mean we query for the previous transaction access mode first, store it and then set it back after the run?

Reply:
How about we set it right back to `READ WRITE` after we are done with this query?

Reply:
I like the `bracket`-like abstraction because it’s less stateful (it just parameterizes a region with dynamic extent, a la environment variables), but you might be right that in this case we really only want to set the access mode for a whole query.

Do you know if there’s any performance cost to issuing `SET TRANSACTION READ WRITE` in a query that’s already in read/write mode? (I feel like there must not be, since it doesn’t actually do anything, but I’m always a little wary of anything that modifies the transaction state.)

Reply:
There is some cost to `SET TRANSACTION READ WRITE` even if it's already in `READ WRITE` mode but I think its practically nothing especially for a metadata API.

I ran a quick test using this function once with `SET TRANSACTION READ WRITE` and next without it

```
CREATE FUNCTION check_transaction(batchsize INTEGER)
RETURNS INTEGER AS $$
DECLARE
   counter INTEGER := 0 ; 
BEGIN
  LOOP
     EXIT WHEN counter = batchsize ;
     counter := counter + 1;
     SET TRANSACTION READ WRITE; // toggle this
  END LOOP;
RETURN counter;
END;
$$  LANGUAGE plpgsql;
```

Results on Heroku hobby-dev:

```
dbuiv0sje6a6v1=> explain analyze select check_transaction(1000);
                                     QUERY PLAN                                     
------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=4.117..4.117 rows=1 loops=1)
 Planning Time: 0.028 ms
 Execution Time: 4.184 ms
dbuiv0sje6a6v1=> explain analyze select check_transaction2(1000);
                                     QUERY PLAN                                     
------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=3.917..3.918 rows=1 loops=1)
 Planning Time: 0.014 ms
 Execution Time: 3.932 ms
(3 rows)

dbuiv0sje6a6v1=> explain analyze select check_transaction(10000);
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=49.254..49.254 rows=1 loops=1)
 Planning Time: 0.010 ms
 Execution Time: 49.266 ms
(3 rows)

dbuiv0sje6a6v1=> explain analyze select check_transaction2(10000);
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=107.129..107.129 rows=1 loops=1)
 Planning Time: 0.014 ms
 Execution Time: 107.196 ms
(3 rows)

dbuiv0sje6a6v1=> explain analyze select check_transaction(100000);
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=508.105..508.106 rows=1 loops=1)
 Planning Time: 0.035 ms
 Execution Time: 508.121 ms
(3 rows)

dbuiv0sje6a6v1=> explain analyze select check_transaction2(100000);
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Result  (cost=0.00..0.05 rows=1 width=4) (actual time=773.483..773.483 rows=1 loops=1)
 Planning Time: 0.013 ms
 Execution Time: 773.500 ms
(3 rows)

Reply:
Sounds good, thanks for checking!